crates/metadata-server/src/grpc/client.rs

```diff
@@ -118,75 +108,68 @@ impl GrpcMetadataServerClient {
         svc_client_guard.clone()
     }
 
-    fn handle_grpc_response<T, E>(
-        &self,
-        response: Result<T, Status>,
-        map_status: impl Fn(Status) -> E,
-    ) -> Result<T, E>
-    where
-        E: MaybeRetryableError,
-    {
-        response.map_err(|status| {
-            let known_leader = KnownLeader::from_status(&status);
-            let err = map_status(status);
-
-            if let Some(known_leader) = known_leader {
-                self.choose_known_leader(known_leader);
-            } else if err.retryable() {
-                self.choose_random_endpoint();
-            }
-            err
-        })
+    fn has_known_leader(&self, status: &Status) -> bool {
+        if let Some(known_leader) = KnownLeader::from_status(status) {
+            self.choose_known_leader(known_leader);
+            true
+        } else {
+            self.choose_random_endpoint();
+            false
+        }
     }
 }
 
 #[async_trait]
 impl MetadataStore for GrpcMetadataServerClient {
     async fn get(&self, key: ByteString) -> Result<Option<VersionedValue>, ReadError> {
-        let retry_policy = Self::retry_policy();
-
-        let response = retry_on_retryable_error(retry_policy, || async {
+        loop {
             let mut client = self
                 .current_client()
                 .ok_or_else(|| ReadError::terminal(NoKnownMetadataServer))?;
 
-            self.handle_grpc_response(
-                client
-                    .get(GetRequest {
-                        key: key.clone().into(),
-                    })
-                    .await,
-                map_status_to_read_error,
-            )
-        })
-        .await?;
-
-        response
-            .into_inner()
-            .try_into()
-            .map_err(|err: ConversionError| ReadError::terminal(err))
+            match client
+                .get(GetRequest {
+                    key: key.clone().into(),
+                })
+                .await
+            {
+                Ok(response) => {
+                    return response
+                        .into_inner()
+                        .try_into()
+                        .map_err(|err: ConversionError| ReadError::terminal(err));
+                }
+                Err(status) => {
+                    // try again if the error response contains information about the known leader
```

Comment:
Can we log here? additionally, can we limit the number of iterations in that loop? I'm okay if we leave this without backoff but this needs to be limited to a very small number of attempts (3?)

Reply:
Makes sense. Will add it.